src/inference/dream/sample.js

```diff
@@ -0,0 +1,152 @@
+    // fantasized data.
+    var trace = new Trace(this.wpplFn, s, env.exit, a);
+    this.record = {trace: trace, data: []};
+
+    this.guideRequired = true;
+    this.insideMapData = false;
+
+    this.coroutine = env.coroutine;
+    env.coroutine = this;
+  }
+
+  dreamSample.prototype = {
+
+    run: function() {
+      return this.wpplFn(_.clone(this.s), function(s, val) {
+        env.coroutine = this.coroutine;
+        return this.cont(this.record);
+      }.bind(this), this.a);
+    },
+
+    sample: function(s, k, a, dist, options) {
+      var sampleFn = this.insideMapData ? this.sampleLocal : this.sampleGlobal;
+      return sampleFn.call(this, s, a, dist, options, function(s, val) {
+        this.record.trace.addChoice(dist, val, a, s, k, options);
+        return k(s, val);
+      }.bind(this));
+    },
+
+    sampleLocal: function(s, a, targetDist, options, k) {
+      return k(s, targetDist.sample());
+    },
+
+    sampleGlobal: function(s, a, dist, options, k) {
+      options = options || {};
+      return guide.getDist(
+          options.guide, options.noAutoGuide, dist, env, s, a,
+          function(s, guideDist) {
+            if (!guideDist) {
+              throw new Error('dream: No guide distribution specified.');
+            }
+            return k(s, guideDist.sample());
+          });
+    },
+
+    factor: function(s, k, a) {
+      throw new Error('dream: factor not supported, use observe instead.');
+    },
+
+    observe: function(s, k, a, dist) {
+      if (!this.insideMapData) {
+        throw new Error('dream: observe can only be used within mapData with this estimator.');
+      }
+      if (!this.obsArr && this.obs.length !== 0) {
+        throw new Error('dream: Expected to see only a single observe per data point.');
+      }
+
+      var val = dist.sample();
+      this.obs.push(val);
+      return k(s, val);
+    },
+
+    mapDataEnter: function() {
+      this.obs = [];
+    },
+
+    mapDataLeave: function() {
+      var datum = this.obsArr ? this.obs : this.obs[0];
+      this.record.data.push(datum);
+    },
+
+    mapDataFetch: function(data, opts, a) {
+      if (this.insideMapData) {
+        throw new Error('dream: nested mapData is not supported by this estimator.');
+      }
+      this.insideMapData = true;
+
+      var batchSize = _.has(opts, 'dreamBatchSize') ? opts.dreamBatchSize : 1;
+      if (!(util.isInteger(batchSize) && batchSize >= 0)) {
+        throw new Error('dream: dreamBatchSize should be a non negative integer.');
+      }
+
+      if (_.isEmpty(data)) {
+        // In general, we need data to yield to the observation
+        // function in order to generate fantasy data.
+        throw new Error('dream: data should be non empty.');
+      }
+
+      // Flag indicating whether each element of the original data is
+      // an array of observations or a single observation. (We check
+      // the first datum, and assume the rest of data would return the
+      // same.)
```

Comment:
You're right. I've [re-worked](https://github.com/null-a/webppl/blob/a57f6fd210ad189e614064cdcc478d672f42a4bd/src/inference/dream/estimator.js#L10) this to unwrap single observations based on the number of times `observe` is called (in a particular application of the observation function) rather than by peeking at the data. I think this is an improvement. I've also added some [tests](https://github.com/probmods/webppl/pull/775/commits/24d447a109f6e5583d21bac89ed45b00acee7cb1) for this stuff.